### PR TITLE
Sample details refactor / metadata tab

### PR DESCRIPTION
**Pull request by deepsidhu85**
## Description of changes
What did you change in this pull request?  Provide a description of files changed, user interactions changed, etc.  Include how to test your changes.

Updated sample metadata tab to allow adding, editing, and deleting sample metadata field and entries.

## Related issue
Link to the GitHub issue this pull request addresses using the `#issuenum` format.  If it completes an issue, use `Fixes #issuenum` to automatically close the issue.

## Checklist
Things for the developer to confirm they've done before the PR should be accepted:

~* [ ] CHANGELOG.md (and UPGRADING.md if necessary) updated with information for new change.~
~* [ ] Tests added (or description of how to test) for any new features.~
~* [ ] User documentation updated for UI or technical changes.~

### CODE REVIEW

File: src/main/java/ca/corefacility/bioinformatics/irida/service/impl/sample/MetadataTemplateServiceImpl.java [hunk 1]
```diff
@@ -155,6 +155,18 @@ public MetadataTemplateField saveMetadataField(MetadataTemplateField field) {
 		return fieldRepository.save(field);
 	}
 
+	@PreAuthorize("permitAll()")
```
Comment (joshsadam) [inline src/main/java/ca/corefacility/bioinformatics/irida/service/impl/sample/MetadataTemplateServiceImpl.java:158]: Do we really want these set as permit all?

File: src/main/webapp/resources/js/apis/samples/samples.js [hunk 2]
```diff
@@ -31,14 +34,77 @@ export const sampleApi = createApi({
...
+  useRemoveSampleMetadataMutation,
+  useUpdateSampleMetadataMutation,
 } = sampleApi;
 
+/**
+ * Gets the sample metadata.
+ * @param {Object} params
+ * @returns {Promise<{}|T>}
+ */
+export const fetchMetadataForSample = async ({ sampleId, projectId }) => {
```
Comment (joshsadam) [inline src/main/webapp/resources/js/apis/samples/samples.js:97]: Shouldn't this be part of the API for a sample?  With a mutation that triggers the update?

**Review by joshsadam — CHANGES_REQUESTED** [review]
Only a few things, this looks really really good.  I will talk to you about 1 other though.

File: src/main/webapp/resources/js/components/samples/components/SampleInfo.jsx [hunk 3]
```diff
@@ -23,12 +23,15 @@ const StyledList = styled(List)`
...
  * @constructor
  */
-export function SampleInfo({ sample, isModifiable }) {
+export function SampleInfo() {
+  const { sample, modifiable: isModifiable } = useSelector(
+    (state) => state.sampleReducer
+  );
+
   const [updateSampleDetails] = useUpdateSampleDetailsMutation();
   const dateFormat = "YYYY-MM-DD";
```
[the remark this answers is not in the document]
Comment (deepsidhu85) [reply src/main/webapp/resources/js/components/samples/components/SampleInfo.jsx:36]: As discussed tried switching from using moment to dayjs via the ant design guide and it does not work. Leaving as is for now

File: src/main/webapp/resources/js/apis/samples/samples.js (hunk 2 above)
Comment (deepsidhu85) [reply src/main/webapp/resources/js/apis/samples/samples.js:97]: As discussed in the SampleMetadata.jsx file we are using a selector to access the metadata from the redux store and using a useeffect hook to update the state. Having this as a part of the sample API resulted in the metadata always returning an empty array, so just followed how we have handled stuff like this in the past

File: src/main/java/ca/corefacility/bioinformatics/irida/service/impl/sample/MetadataTemplateServiceImpl.java (hunk 1 above)
Comment (deepsidhu85) [reply src/main/java/ca/corefacility/bioinformatics/irida/service/impl/sample/MetadataTemplateServiceImpl.java:158]: As discussed I will setup a meeting so we can go over these. The permitAll is used for the other metadatatemplatefield methods as well. Leaving as is for now but after the discussion we can see where to go with it in a future pr

File: src/main/webapp/resources/js/pages/analysis/components/settings/AnalysisSampleRenderer.jsx [hunk 4]
```diff
@@ -66,14 +66,11 @@ export function AnalysisSampleRenderer() {
...
-                    <a
-                      href={`${SAMPLES_BASE_URL}/${item.sampleId}/details`}
-                      target="_blank"
-                      className="t-paired-end-sample-name"
-                      style={{ color: blue6 }}
-                    >
-                      {item.sampleName}
-                    </a>
+                    <SampleDetailViewer sampleId={item.sampleId}>
+                      <Button style={{ padding: "0px" }} type="link">
```
[the remark this answers is not in the document]
Comment (deepsidhu85) [reply src/main/webapp/resources/js/pages/analysis/components/settings/AnalysisSampleRenderer.jsx]: Removed this for now and reverted to using a link to the sample details page for now. We will need to discuss this as when we access the sample from an analyses page there is no associated project

File: src/main/webapp/resources/js/components/samples/store.js [hunk 5]
```diff
@@ -1,8 +1,10 @@
 import { configureStore } from "@reduxjs/toolkit";
 import { sampleApi } from "../../apis/samples/samples";
+import sampleReducer from "./sampleSlice";
 
 export default configureStore({
```
[the remark this answers is not in the document]
Comment (deepsidhu85) [reply src/main/webapp/resources/js/components/samples/store.js:9]: Added in 5edbd2a28ddf96c80f7b2cd6f768b60a34fb4c4d

File: src/main/webapp/resources/js/components/samples/components/EditMetadata.jsx [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+        projectId,
+        metadataFieldId: fieldId,
+        metadataField: values.metadata_field_name,
+        metadataEntryId: entryId,
+        metadataEntry: values.metadata_field_value,
+        metadataRestriction: values.metadata_field_permission,
+      })
+        .then((response) => {
+          if (response.error) {
+            notification.error({ message: response.error.data.error });
```
Comment (joshsadam) [inline src/main/webapp/resources/js/components/samples/components/EditMetadata.jsx:61]: Does this ever get hit?  Should it not get caught on line 77?

Comment (deepsidhu85) [reply src/main/webapp/resources/js/components/samples/components/EditMetadata.jsx]: You are correct. Looks like I forgot to remove that once I moved the catching of it into the then statement. Removed in 54adb73843ab4d44d03e431f3dc155e70ea2b16d

**Review by joshsadam — APPROVED** [review]
This looks good.  Do you have any other changes you would like to make?

**Pull request merged**